### ISSUE TITLE
Mean Absolute Score : Overall Importance

### ISSUE THREAD

**Issue by Tejamr** [interpretml/interpret#337]
Hi , Can I know how Mean absolute score for each feature(Feature Importance) is calculated in EBM global explanations?
Is there any particular metric or mathematical formula to calculate those probabilities? If, yes please let me know .
Please check below png . May I know how those values are calculated . How Glucose has got more score irrespective of others ?
![newplot](https://user-images.githubusercontent.com/91931018/167353198-221f02d0-3215-4b23-a88f-b083d514092d.png)

**Comment by Tejamr** [comment]
Can you please explain by taking any dataset or values as example?How the probability values are coding?

**Comment by Harsha-Nori** [comment]
Hi @Tejamr, 

I think the simplest setting is a boolean feature with a regression dataset. Let's say we're predicting housing price, and the feature of interest is "A/C" (so 1 means the house has air conditioning, 0 means it doesn't). In the training dataset, we have 1,000 houses with A/C and 2,000 without A/C. Let's also assume the model learns -$10,000 for no A/C and +$15,000 for A/C houses.

So to summarize:

Counts: {No A/C: 2,000,  A/C: 1,000}
Model Scores: {No A/C: -$10,000, A/C: +$15,000}

These values are the same ones you can see for the feature when you call `show(ebm.explain_global())` -- the counts are in the density graph at the bottom, and the model scores are in the main figure. Make sure you are inspecting the learned functions for your features too by using the dropdown in explain_global()!

The overall feature importance for the "A/C?" feature is simply the average absolute contribution that feature makes on each sample in the training set. So in this case, it would be:

> (2,000 * abs(-$10,000) + 1,000 * abs($15,000))  /  (2,000 + 1,000)    

> (2,000 * $10,000 + 1,000 * $15,000) / (2,000 + 1,000)

> 35,000,000 / 3,000 = $11,666

The way to interpret this is that on average, the "A/C?" feature moved predictions _in some direction_ by $11,666. Intuitively this makes sense too -- for every sample, it either reduced the price by $10,000 or increased it by $15,000, so the average importance must be somewhere in between. In this case, it's closer to $10,000 because the majority of the houses do not have A/C.

It's the exact same procedure for classification, only the scores are in log odds instead of in the units of the target. For continuous features or categorical features with more than 2 attributes, it's the same procedure again, except we have more terms in the weighted average calculation (one for each bin of data in the feature).

Note that this isn't the only way of calculating feature importance, and none of them are perfect. For example, this method will not highlight features that learn incredibly strong effects that effect a tiny number of samples. Other metrics can be the max() - min() of the graph, the change in RMSE from removing a feature, etc. -- we just picked one that seemed reasonably broadly applicable. Happy to clarify further if you have any other questions!

**Comment by Tejamr** [comment]
Hi @Harsha-Nori ,

Thanks for your detailed explanation . But still a bit confusing for me regarding mathematical calculations of Model scores . I am working on PIMA Diabetes dataset(Kaggle) and here is the dataset . I have applied EBM on this dataset and got global explanations as below .

Can u please explain me the detailed calculations like how probability scores are calculate for each feature .
Ex : How Glucose has got more score regrads of other features . When I plot correlation Pregnancies has more impact on the dependant feature but in EBM Global explanations pedigree function contributing more , How this is happening?

A detailed mathematical calculations will be more helpful as I am supposed to submit POC .
![newplot (1)](https://user-images.githubusercontent.com/91931018/167774457-574c1923-f09a-4a2c-8181-0aa627bb96c3.png)

[diabetes.csv](https://github.com/interpretml/interpret/files/8666919/diabetes.csv)

**Comment by Tejamr** [comment]
Hi @Harsha-Nori ,@interpret-ml ,

Still same confusion fr me . I am not able to get the probability values as I am a noob . Can u please calculate those values and share the notebook or handwritten paper here. I mean I need Calculations from the scratch like:
1:How additive terms are calculated?
2.How probability values are calculated?
3.How values are passed to the shape function and how they will be converted to logit link functions?
4.How we get each and every value .I am asking like internal structure of EBM with mathematical calculations on diabetes dataset .

Heres the dataset for your reference .

[diabetes.csv](https://github.com/interpretml/interpret/files/8675198/diabetes.csv)

I am not at all getting the single point also regarding calculations . It will be a great beneficiary for me if u help me in this .

**Comment by Tejamr** [comment]
Hi @Harsha-Nori ,@Interpret-ml

Can u pls explain what is the shape function here used . How should I pass those values to the shape function .

Note : All this means is that we take all of the data in the training set, select only the column of data for the feature we're interested in, pass it through the shape function that EBM learned (which you can visualize with the dropdown in ebm.explain_global(), take the absolute value, and average everything to get the final answer.

*For the above explanation I am asking ...

**Comment by Harsha-Nori** [comment]
Hi @Tejamr, 

No worries. I highly recommend watching the first ~20 minutes of this video: https://www.youtube.com/watch?v=2YKtNYBuojE  where Rich Caruana explains what EBMs and GAMs are, how we use boosted decision trees to learn the graphs, and how we make a single prediction with a graph. 

This chapter on GAMs from the interpretable machine learning book might be helpful too: https://christophm.github.io/interpretable-ml-book/extend-lm.html . I truly think the explanations above will make more sense to you once you've understood how GAMs and other additive models work!

**Comment by Tejamr** [comment]
Hi @Harsha-Nori ,

Thank you for your detailed explanation . But, can I know how we are getting additive terms per each feature .
Ex : ebm.additive_terms_per feature_
above line is giving a set of arrays with some bunch of point values . When I gone through the code it was like averaging the absolute additive terms will give the overall model feature importances(Global).Can I know how those additive terms are coming . Is there any particular method or mathematical calculation behind that?If yes, please let me know .

**Comment by Tejamr** [comment]
Hi @Harsha-Nori ,

In the above article u have mentioned like we have to take a single column value and pass it through the shape function which we have defined . U have defined a formula also .I am using EBM algorithm can I know what is the shape function for this and what is the formula for the shape function we have to pass .Kindly allow me to know abt these statistics .
